File: R/TranslocRepeatSeq.R
#!/usr/bin/env Rscript

if (commandArgs()[1] != "RStudio") {
  
  ARGS <- c(
    "tlxfile", "character", "",
    "bedfile","character","",
    "output","character", ""
  )
  
  OPTS <- c(
  
  )

  source_local <- function(fname){
    argv <- commandArgs(trailingOnly = FALSE)
    base_dir <- dirname(substring(argv[grep("--file=", argv)], 8))
    source(paste(base_dir, fname, sep="/"))
  }
  
  source_local("Rsub.R")
  parseArgs("TranslocRepeatSeq.R", ARGS, OPTS)
  
} else {
  source("~/TranslocPipeline/R/Rsub.R")
  source("~/TranslocPipeline/R/TranslocHelper.R")
  tlxfile <- "./VK020_Alt133.tlx"
  output <- "./VK020_Alt133_repeatseq.txt"
  bedfile <- "/Volumes/AltLab/Genomes/mm9/annotation/repeatSeq.bed"
}

suppressPackageStartupMessages(library(data.table, quietly=TRUE))
suppressPackageStartupMessages(library(dplyr, quietly=TRUE))
suppressPackageStartupMessages(library(GenomicRanges, quietly=TRUE))
suppressPackageStartupMessages(library(rtracklayer, quietly=TRUE))

bed <- import.bed(bedfile)

tlx <- fread(tlxfile,sep="\t",header=T,select=c("Qname","JuncID","Rname","Junction"))

tlx <- filter(tlx, ! is.na(Rname) & ! is.na(Junction))

gr <-  with(tlx,GRanges(seqnames=Rname,ranges=IRanges(start=Junction,width=1,names=Qname)))

ol <- suppressWarnings(overlapsAny(gr,bed,type="any"))

tlx <- mutate(tlx, Overlap = ol) %>% filter(Overlap == T) %>% select(Qname,JuncID)

write.table(tlx,output,sep="\t",col.names=F,row.names=F,quote=F,na="")

File: bin/TranslocRepeatSeq.pl
#!/usr/bin/env perl


use strict;
use warnings;
use Getopt::Long;
use Carp;
use IO::File;
use Text::CSV;
use Interpolation 'arg:@->$' => \&argument;
use Time::HiRes qw(gettimeofday tv_interval);

use Cwd qw(abs_path);
use FindBin;
use lib abs_path("$FindBin::Bin/../lib");


require "TranslocHelper.pl";
require "PerlSub.pl";

# Flush output after every write
select( (select(STDOUT), $| = 1 )[0] );

##
## This program r
## run with "--help" for usage information
##
## Robin Meyers

# Forward declarations
sub parse_command_line;

# Global flags and arguments, 
# Set by command line arguments
my $tlxfile;
my $bedfile;
my $output;


# Global variabless
my %filtered_reads;
my $repeatseq_output;
#
# Start of Program
#

parse_command_line;

# Run Rscript to print list of junctions to filter
($repeatseq_output = $output) =~ s/\.tlx$/.txt/;

my $repeatseq_cmd = join(" ","$FindBin::Bin/../R/TranslocRepeatSeq.R",
                        $tlxfile,
                        $bedfile,
                        $repeatseq_output);

System($repeatseq_cmd);

# Read in filtered junctions
open JUNC, "<", $repeatseq_output;
while (<JUNC>) {
  chomp;
  my @read = split("\t");
  $filtered_reads{$read[0]}->{$read[1]} = 1;
}
close JUNC;


# Read in tlxfile line by line, writing to output

my $infh = IO::File->new("<$tlxfile");
my $outfh = IO::File->new(">$output");
my $csv = Text::CSV->new({sep_char => "\t"});
my $header = $csv->getline($infh);
$csv->column_names(@$header);

$outfh->print(join("\t",@$header)."\n");

my $junc_id;
my $qname;

while (my $tlx = $csv->getline_hr($infh)) {

  if (! defined $qname || $tlx->{Qname} ne $qname) {
    $junc_id = 1;
    $qname = $tlx->{Qname}
  } else {
    $junc_id++;
  }

  if (defined $filtered_reads{$tlx->{Qname}}->{$junc_id}) {
    $tlx->{repeatseq} = 1;
    print "found one\n";
  }
  $outfh->print(join("\t",@{$tlx}{@$header})."\n");
}


#
# End of program
#


sub parse_command_line {

  my $help;

  usage() if (scalar @ARGV == 0);

  my $result = GetOptions ( 
                            "help" => \$help

                          );
  
  usage() if ($help);



  #Check options

  usage() if (scalar @ARGV < 3);

  $tlxfile = shift(@ARGV);
  $bedfile = shift(@ARGV);
  $output = shift(@ARGV);

  croak "Error: cannot read tlxfile" unless -r $tlxfile;
  croak "Error: cannot read bedfile" unless -r $bedfile;
  croak "Error: tlxfile must have .tlx extension" unless $tlxfile =~ /\.tlx$/;


  exit unless $result;
}


sub usage()
{
print<<EOF;
TranslocRepeatSeq, by Robin Meyers, 2013

Usage: $0 tlxfile bedfile output
        [--option VAL] [--flag] [--help]


Arguments (defaults in parentheses):

$arg{"tlxfile","File containing meta data for one experiment per row - follow correct format"}
$arg{"bedfile","File containing meta data for one experiment per row - follow correct format"}
$arg{"output","File containing meta data for one experiment per row - follow correct format"}

$arg{"--help","This helpful help screen."}

--------------------------------------------

EOF

exit 1;
}
